File: ohara-agent/src/test/scala/com/island/ohara/agent/SshdServer.scala
package com.island.ohara.agent

import java.io.{InputStream, OutputStream}
import java.util

import com.island.ohara.common.util.{CloseOnce, CommonUtil}
import org.apache.sshd.server.command.Command
import org.apache.sshd.server.keyprovider.SimpleGeneratorHostKeyProvider
import org.apache.sshd.server.shell.{ProcessShellCommandFactory, ProcessShellFactory}
import org.apache.sshd.server.{Environment, ExitCallback, SshServer}

/**
  * a embedded ssh server.
  */
trait SshdServer extends CloseOnce {

  /**
    * @return ssh server's hostname
    */
  def hostname: String

  /**
    * @return ssh server's port
    */
  def port: Int

  /**
    * @return ssh client's user
    */
  def user: String

  /**
    * @return ssh client's password
    */
  def password: String
}

object SshdServer {

  /**
    * used to handle the command manually. It is useful when the command you want to test doesn't work on local machine.
    */
  trait CommandHandler {

    /**
      * @param command will be executed on ssh server
      * @return true if this handler want to handle the command
      */
    def belong(command: String): Boolean

    /**
      * @param command will be executed on ssh server
      * @return response of the command
      */
    def execute(command: String): Seq[String]
  }

  private[agent] val SSHD_SERVER: String = "ohara.it.sshd"

  private[agent] def parseString(sshdString: String): (String, String, String, Int) = {
    // format => user:password@host:port
    try {
      val user = sshdString.split(":")(0)
      val password = sshdString.split(":")(1).split("@").head
      val host = sshdString.split(":")(1).split("@").last
      val port = sshdString.split(":").last.toInt
      (user, password, host, port)
    } catch {
      case e: Throwable => throw new IllegalArgumentException(s"invalid value from $SSHD_SERVER", e)
    }
  }

  def apply(): SshdServer = apply(sys.env.get(SSHD_SERVER))

  private[agent] def apply(sshdString: Option[String]): SshdServer = sshdString
    .map { s =>
      val (_user, _password, _hostname, _port) = parseString(s)
      new SshdServer {
        override def hostname: String = _hostname
        override def port: Int = _port
        override def user: String = _user
        override def password: String = _password
        override protected def doClose(): Unit = {}
      }
    }
    .getOrElse(local(0))

  /**
    * create a embedded sshd server bound on specified port
    * @param port listen port. If the port is smaller than zero, a random port is used in running sshd server
    * @param commandHandlers used in testing. Control the response for specified command. for example.
    *                       Map("hello" -> "world") means the command "hello" will get the response "world"
    * @return a embedded sshd server
    */
  def local(port: Int, commandHandlers: Seq[CommandHandler] = Seq.empty): SshdServer = {
    val _user = CommonUtil.uuid()
    val _password = CommonUtil.uuid()
    val sshd = SshServer.setUpDefaultServer()
    sshd.setKeyPairProvider(new SimpleGeneratorHostKeyProvider())
    sshd.setPasswordAuthenticator((username, password, _) => username == _user && password == _password)
    sshd.setShellFactory(new ProcessShellFactory(util.Arrays.asList("/bin/sh", "-i", "-l")))
    sshd.setCommandFactory(command => {
      commandHandlers
        .find(h => h.belong(command))
        .map {
          handler =>
            new Command {
              private[this] var out: OutputStream = _
              private[this] var callback: ExitCallback = _
              override def setInputStream(in: InputStream): Unit = {}

              override def setOutputStream(out: OutputStream): Unit = this.out = out

              override def setErrorStream(err: OutputStream): Unit = {}

              override def setExitCallback(callback: ExitCallback): Unit = this.callback = callback

              override def start(env: Environment): Unit = {
                handler.execute(command).foreach { s =>
                  out.write(s.getBytes)
                  // TODO: make it configurable...by chia
                  out.write("\n".getBytes)
                }
                callback.onExit(0)
              }
              override def destroy(): Unit = {}
            }

        }
        .getOrElse(ProcessShellCommandFactory.INSTANCE.createCommand(command))
    })
    sshd.setHost(CommonUtil.hostname())
    sshd.setPort(if (port <= 0) 0 else port)
    sshd.start()
    new SshdServer {
      override def hostname: String = sshd.getHost
      override def port: Int = sshd.getPort
      override def user: String = _user
      override def password: String = _password
      override protected def doClose(): Unit = sshd.close()
    }
  }

}

File: ohara-testing-util/src/main/java/com/island/ohara/testing/service/SshdServer.java
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.testing.service;

import com.island.ohara.common.util.CommonUtils;
import com.island.ohara.common.util.Releasable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.apache.sshd.server.Environment;
import org.apache.sshd.server.ExitCallback;
import org.apache.sshd.server.SshServer;
import org.apache.sshd.server.command.Command;
import org.apache.sshd.server.keyprovider.SimpleGeneratorHostKeyProvider;
import org.apache.sshd.server.session.ServerSession;
import org.apache.sshd.server.shell.ProcessShellCommandFactory;
import org.apache.sshd.server.shell.ProcessShellFactory;

public interface SshdServer extends Releasable {

  /** @return ssh server's hostname */
  String hostname();

  /** @return ssh server's port */
  int port();

  /** @return ssh client's user */
  String user();

  /** @return ssh client's password */
  String password();

  interface CommandHandler {
    /**
     * @param cmd will be executed on ssh server
     * @return true if this handler want to handle the command
     */
    boolean belong(String cmd);

    /**
     * @param cmd will be executed on ssh server
     * @return response of the command
     */
    List<String> execute(String cmd);
  }

  static SshdServer local() {
    return local(0);
  }

  static SshdServer local(int port) {
    return local(port, Collections.emptyList());
  }

  static SshdServer local(int port, List<CommandHandler> handlers) {
    String _user = CommonUtils.randomString();
    String _password = CommonUtils.randomString();
    SshServer sshd = SshServer.setUpDefaultServer();
    sshd.setKeyPairProvider(new SimpleGeneratorHostKeyProvider());
    sshd.setPasswordAuthenticator(
        (String username, String password, ServerSession session) ->
            username.equals(_user) && password.equals(_password));
    sshd.setShellFactory(new ProcessShellFactory(Arrays.asList("/bin/sh", "-i", "-l")));
    sshd.setCommandFactory(
        (String command) ->
            handlers
                .stream()
                .filter(h -> h.belong(command))
                .findFirst()
                .map(
                    h ->
                        (Command)
                            new Command() {
                              private OutputStream out = null;
                              private OutputStream err = null;
                              private ExitCallback callback = null;

                              @Override
                              public void start(Environment env) {
                                try {
                                  h.execute(command)
                                      .forEach(
                                          s -> {
                                            try {
                                              out.write(s.getBytes());
                                              // TODO: make it configurable...by chia
                                              out.write("\n".getBytes());
                                            } catch (Throwable e) {
                                              throw new RuntimeException(e);
                                            }
                                          });
                                  callback.onExit(0);
                                } catch (Throwable e) {
                                  callback.onExit(2, e.getMessage());
                                }
                              }

                              @Override
                              public void destroy() {
                                Releasable.close(out);
                                Releasable.close(err);
                              }

                              @Override
                              public void setInputStream(InputStream in) {
                                // do nothing
                              }

                              @Override
                              public void setOutputStream(OutputStream out) {
                                this.out = out;
                              }

                              @Override
                              public void setErrorStream(OutputStream err) {
                                this.err = err;
                              }

                              @Override
                              public void setExitCallback(ExitCallback callback) {
                                this.callback = callback;
                              }
                            })
                .orElseGet(() -> ProcessShellCommandFactory.INSTANCE.createCommand(command)));
    sshd.setHost(CommonUtils.hostname());
    sshd.setPort(Math.max(port, 0));
    try {
      sshd.start();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return new SshdServer() {

      @Override
      public void close() {
        Releasable.close(sshd);
      }

      @Override
      public String hostname() {
        return sshd.getHost();
      }

      @Override
      public int port() {
        return sshd.getPort();
      }

      @Override
      public String user() {
        return _user;
      }

      @Override
      public String password() {
        return _password;
      }
    };
  }
}
